algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala

```diff
@@ -1027,16 +1082,58 @@ case class TopNCMSAggregator[K](cmsMonoid: TopNCMSMonoid[K])
  * family of the form:
  *
  * `h(x) = [a * x + b (mod p)] (mod m)`
+ *
+ * As a requirement for using CMS you must provide an implicit `CMSHasher[K]` for the type `K` of the items you want to
+ * count.  Algebird ships with several such implicits for commonly used types `K` such as [[Long]] and [[BigInt]]:
+ *
+ * {{{
+ * import com.twitter.algebird.CMSHasherImplicits._
+ * }}}
+ *
+ * If your type `K` is not supported out of the box, you have two options: 1) You provide a "translation" function to
+ * convert items of your (unsupported) type `K` to a supported type such as `Array[Byte]`, and then use the `contramap`
+ * function of [[CMSHasher]] to create the required `CMSHasher[K]` for your type (see the documentation of `contramap`
+ * for an example); 2) You implement a `CMSHasher[K]` from scratch, using the existing CMSHasher implementations as a
+ * starting point.
  */
 trait CMSHasher[K] extends java.io.Serializable {
 
-  val PRIME_MODULUS = Int.MaxValue
-
   /**
    * Returns `a * x + b (mod p) (mod width)`.
    */
   def hash(a: Int, b: Int, width: Int)(x: K): Int
 
+  /**
+   * Given `f`, a function from `L` into `K`, creates a `CMSHasher[L]` whose hash function is equivalent to:
+   *
+   * {{{
+   * def hash(a: Int, b: Int, width: Int)(x: L): CMSHasher[L] = CMSHasher[K].hash(a, b, width)(f(x))
+   * }}}
+   */
+  def on[L](f: L => K)(implicit hasher: CMSHasher[K]) = new CMSHasher[L] {
```

Comment:
why do we need the implicit hasher here rather than doing `self.hash(a, b, width)(f(x))` below? This method, with the implicit parameter, seems appropriate for the `object` but not the trait. I don't see where this method uses `this` or `self`, which is what makes me think something should change.